qiskit/quantum_info/operators/symplectic/clifford_circuits.py

```diff
@@ -39,10 +39,44 @@ def _append_circuit(clifford, circuit, qargs=None):
     if qargs is None:
         qargs = list(range(clifford.num_qubits))
 
+    if isinstance(circuit, Instruction):
+        return _append_operation(clifford, circuit, qargs)
     if isinstance(circuit, QuantumCircuit):
-        gate = circuit.to_instruction()
-    else:
-        gate = circuit
+        qubit_indices = {bit: idx for idx, bit in enumerate(circuit.qubits)}
+        for instruction in circuit:
+            if instruction.clbits:
+                raise QiskitError(
+                    f"Cannot apply Instruction with classical bits: {instruction.operation.name}"
+                )
+            # Get the integer position of the flat register
+            new_qubits = [qargs[qubit_indices[tup]] for tup in instruction.qubits]
+            _append_operation(clifford, instruction.operation, new_qubits)
+        return clifford
+
+    raise QiskitError("The circuit must be QuantumCircuit or Instruction")
+
+
+def _append_operation(clifford, operation, qargs=None):
+    """Update Clifford inplace by applying a Clifford operation.
+
+    Args:
+        clifford (Clifford): the Clifford to update.
+        operation (Instruction or str): the operation or composite operation to apply.
+        qargs (list or None): The qubits to apply operation to.
+
+    Returns:
+        Clifford: the updated Clifford.
+
+    Raises:
+        QiskitError: if input operation cannot be decomposed into Clifford operations.
+    """
+    if isinstance(operation, (Barrier, Delay)):
+        return clifford
+
+    if qargs is None:
+        qargs = list(range(clifford.num_qubits))
+
+    gate = operation
 
     # Basis Clifford Gates
     basis_1q = {
```

Comment:
If you need more speed, you might want to refactor the `basis_1q` and `basis_2q` lookups into a global hashmap so they only need to be constructed once.  It probably won't affect much, but it might be a shade faster.

Reply:
Done in [226892c](https://github.com/Qiskit/qiskit-terra/pull/8585/commits/226892c2bf30e3362a247b593359209dc2b6c2fc)